exes/mqtt-broadcaster/inc/mqtt_broadcaster.hpp

```diff
@@ -0,0 +1,560 @@
+#pragma once
+
+#include <sparkplug_b/sparkplug_b.pb.h>
+#include <async_mqtt/all.hpp>
+#include <boost/asio.hpp>
+
+#include <tfc/dbus/string_maker.hpp>
+
+#include "config.hpp"
+
+using org::eclipse::tahu::protobuf::Payload;
+using org::eclipse::tahu::protobuf::Payload_Metric;
+
+namespace asio = boost::asio;
+namespace details = tfc::ipc::details;
+
+class network_manager {
+public:
+  auto connect_socket(auto&& socket, auto&& resolved_ip) -> asio::awaitable<void> {
+    co_await asio::async_connect(socket, resolved_ip, asio::use_awaitable);
+  }
+
+  auto set_sni_hostname(auto&& native_handle, std::string broker_address) -> void {
+    if (!SSL_set_tlsext_host_name(native_handle, broker_address.c_str())) {
+      boost::system::error_code error{ static_cast<int>(::ERR_get_error()), boost::asio::error::get_ssl_category() };
+
+      throw boost::system::system_error{ error };
+    }
+  }
+
+  auto async_handshake(auto&& socket) -> asio::awaitable<void> {
+    co_await socket.async_handshake(async_mqtt::tls::stream_base::client, asio::use_awaitable);
+  }
+};
+
+struct signal_data {
+  tfc::ipc_ruler::signal information;
+  tfc::ipc::details::any_recv receiver;
+  std::optional<std::any> current_value;
+};
+
+template <class ipc_client_type, class mqtt_client_type, class config_type, class network_manager_type>
+class mqtt_broadcaster {
+public:
+  static constexpr std::string_view namespace_element = "spBv1.0";
+  static constexpr std::string_view ndata = "NDATA";
+  static constexpr std::string_view nbirth = "NBIRTH";
+  static constexpr std::string_view ndeath = "NDEATH";
+  static constexpr std::string_view ncmd = "NCMD";
+  static constexpr std::string_view rebirth_metric = "Node Control/Rebirth";
+
+  explicit mqtt_broadcaster(asio::io_context& io_ctx)
+      : io_ctx_(io_ctx), tls_ctx_{ async_mqtt::tls::context::tlsv12 }, ipc_client_{ io_ctx_ },
+        mqtt_client_(std::make_shared<mqtt_client_type>(async_mqtt::protocol_version::v5, io_ctx.get_executor(), tls_ctx_)),
+        logger_("mqtt_broadcaster"), config_(io_ctx, "mqtt_broadcaster"), network_manager_() {
+    properties_callback_ = ipc_client_.register_properties_change_callback(
+        [this]([[maybe_unused]] sdbusplus::message_t& msg) { add_new_signals(); });
+    asio::co_spawn(mqtt_client_->strand(), initialize(), asio::detached);
+  }
+
+  explicit mqtt_broadcaster(asio::io_context& io_ctx, bool)
+      : io_ctx_(io_ctx), tls_ctx_{ async_mqtt::tls::context::tlsv12 }, ipc_client_{ io_ctx_ },
+        mqtt_client_(std::make_shared<mqtt_client_type>(async_mqtt::protocol_version::v5, io_ctx.get_executor(), tls_ctx_)),
+        logger_("mqtt_broadcaster"), config_(io_ctx, "mqtt_broadcaster"), network_manager_() {}
```

Comment:
what is this constructor doing? It accepts an unused boolean

Reply:
The unused boolean is a subpar solution but the point of this constructor is that unit tests can use it and skip the following lines:

```cpp
properties_callback_ = ipc_client_.register_properties_change_callback([this]([[maybe_unused]]sdbusplus::message_t& msg) { add_new_signals(); });
asio::co_spawn(mqtt_client_->strand(), initialize(), asio::detached);
```

which prevents the class from starting to run the initialization function, allowing the tests to run each function individually.

Reply:
I would recommend either to make a comment on this constructor

Or, remove this constructor, make initialize do everything that is needed or extract new function `run` and call that function within main.

Reply:
> make initialize do everything that is needed

That would just be the normal constructor, which has a lot of side effects from the initialization function.

> extract new function `run` and call that function within main.

If the constructor calls `run` in both testing and normal execution then I still have the same problem.

Reply:
Constructor has been removed and a run() function has been added which does the same job as the constructor along with running the io ctx. The tests would ignore this function when unit testing. That way the callee only needs to call:

```cpp
mqtt_broadcaster<tfc::ipc_ruler::ipc_manager_client, async_mqtt::endpoint<async_mqtt::role::client, async_mqtt::protocol::mqtts>, tfc::confman::config<config>, network_manager> application(io_ctx);

application.run();
```